### DISCUSSION TITLE
Clean partition table from unwanted partitions?

### DISCUSSION THREAD

**Discussion by mickeyproc** [OE4T/meta-tegra#1476] [Q&A] [answered]
I've got myself a Orin Nano dev board with attached nvme, that I've successfully flashed with a custom yocto distro (with a bbappend to tegra-storage-layout-base).
I got Rauc set up with nvbootctrl and UEFI using the described Redundant AB from the Wiki.
The next step would be to completely remove all the partitions that aren't needed.
I know the default partition xml files have tons of partitions with *required* tag in the descriptions, but these are seemingly only for recovery, that i don't need to have and have no use for.
What we need is a clean partition table with only a recovery rootfs, the rootfs A/B, the bootloader partition and a data partition that can be mounted and used by both A/B rootfs once booted up.
So i tried to clean up my custom partition layout xml file from all the "unnecesarry" partitions and build a tegraflash image that could be flashed to the Orin Nano nvme. This fails and goes straight to UEFI shell.

```
<?xml version="1.0"?>
<partition_layout version="01.00.0000">
    <device type="external" instance="0" sector_size="512" num_sectors="NUM_SECTORS">
        <partition name="master_boot_record" type="protective_master_boot_record">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> 512 </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 8 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <description> **Required.** Contains protective MBR. </description>
        </partition>
        <partition name="primary_gpt" type="primary_gpt">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> 19968 </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 8 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <description> **Required.** Contains primary GPT of the `nvme` device. All
              partitions defined after this entry are configured in the kernel, and are
              accessible by standard partition tools such as gdisk and parted. </description>
        </partition>
        <partition name="RECOVERY" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> RECROOTFSSIZE </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x8 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <description> **Optional.** Reserved for future use by the recovery filesystem;
              removable. </description>
        </partition>
        <partition name="APP" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> APPSIZE </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x8 </allocation_attribute>
            <align_boundary> 4096 </align_boundary>
            <percent_reserved> 0 </percent_reserved>
            <filename> APPFILE </filename>
            <unique_guid> APPUUID </unique_guid>
            <description> **Required.** Contains the rootfs. This partition must be defined
              after `primary_GPT` so that it can be accessed as the fixed known special device
              `/dev/nvme0n1p2`. </description>
        </partition>
        <partition name="APP_b" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> APPSIZE </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x8 </allocation_attribute>
            <align_boundary> 4096 </align_boundary>
            <percent_reserved> 0 </percent_reserved>
            <filename> APPFILE_b </filename>
            <unique_guid> APPUUID_b </unique_guid>
            <description> **Required.** Contains the rootfs. This partition must be defined
              after `primary_GPT` so that it can be accessed as the fixed known special device
              `/dev/nvme0n1p3`. </description>
        </partition>
        <partition name="esp" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> 67108864 </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x8 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <filename> ESP_FILE </filename>
            <partition_type_guid> C12A7328-F81F-11D2-BA4B-00A0C93EC93B </partition_type_guid>
            <description> **Required.** EFI system partition with L4T Launcher. </description>
        </partition>
        <partition name="DATA" type="data">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> 18432 </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 0x808 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <description> **Required.** Automatically takes all remaining space on the device except that
              occupied by the `secondary_gpt` partition. Allocation attribute must be set to 0x808.
              May be mounted and used to store user data. </description>
        </partition>
        <partition name="secondary_gpt" type="secondary_gpt">
            <allocation_policy> sequential </allocation_policy>
            <filesystem_type> basic </filesystem_type>
            <size> 0xFFFFFFFFFFFFFFFF </size>
            <file_system_attribute> 0 </file_system_attribute>
            <allocation_attribute> 8 </allocation_attribute>
            <percent_reserved> 0 </percent_reserved>
            <description> **Required.** Contains secondary GPT of the `nvme`
              device. </description>
        </partition>
    </device>
</partition_layout>
```

Has anyone tried cleaning up these partitions or have an idea on how to do this?

UPDATE:
Maybe i should rephrase.
As I understand, the kernel is loaded from the kernel partition instead of from the rootfs (I don't see any kernel image from the rootfs at least). What i want to do, is _only_ have the kernel in the rootfs and not depend on kernel partitions. I basically just want all the partitions to be part of the rootfs instead of split up like this. I know it has benefits to have a partition table like this, but it's not needed for for me.

**Comment by madisongh** [comment]
I've experimented with removing the kernel and kernel-dtb Android partitions in the past, and IIRC it did work, just removing the entries entirely from the storage layout XML file.  You'll probably have to re-flash (including the SPI flash) after making those changes, and make sure you have `UBOOT_EXTLINUX = "1"` set so that the `/boot` tree gets populated.

**Reply by mickeyproc** [reply]
Good news that you've gotten it to work in the past, just by removing the kernel/kernel-dtb android partitions!
I am using the `initrd-flash` script to flash the device and have a serial output hooked up to see what's happening.
I did notice `UBOOT_EXTLINUX = "1"` populated the `/boot` dir. A bit misleading though, that it's called that, as neither uboot nor extlinux is used. I'll try, in correlation with [this discussion](https://github.com/OE4T/meta-tegra/discussions/1480), to also remove the L4TLauncher and just use UEFI (and maybe grub on top). Just to make it cleaner in regards on what's actually being put on the target and what's being used.
Do you have any advice/remarks on this?

**Reply by madisongh** [reply]
>  A bit misleading though, that it's called that, as neither uboot nor extlinux is used.

True, but OE-Core already had support for constructing `extlinux.conf` files based on the `UBOOT_EXTLINUX` variables, and L4TLauncher is  emulating U-Boot's extlinux support, so it seemed like a good idea to just reuse what was already there.

>I'll try, in correlation with https://github.com/OE4T/meta-tegra/discussions/1480, to also remove the L4TLauncher and just use UEFI (and maybe grub on top). Just to make it cleaner in regards on what's actually being put on the target and what's being used.
Do you have any advice/remarks on this?

Hopefully it will work.  Last time I tried it (using systemd-boot as the loader instead of grub), I didn't get anywhere - UEFI refused to load my `bootaa64.efi` and I couldn't coax it into telling me why.

**Reply by ricardosalveti** [reply]
extlinux.conf works fine and also what we use with ostree (but we add an additional patch to make it work with syslinux instead, as that is what ostree produces - https://github.com/OE4T/meta-tegra/blob/kirkstone/external/sota/recipes-bsp/uefi/edk2-firmware-tegra/0001-L4TLauncher-boot-syslinux-instead-of-extlinux-for-os.patch).

**Reply by mickeyproc** [reply]
Sorry the long time since response @madisongh and @ricardosalveti
I've created a `.bbappend` to `tegra-binaries` recipe and added my own partition layout.
Then I added `UBOOT_EXTLINUX = "1"` to my image recipe and also had to add `l4t-launcher-extlinux` to my `IMAGE_INSTALL`.
Then after using calling `./initrd-flash` from the resulting tegraflash binary, i got the device to boot up using the new partition layout.

The problem is now that the new DATA partition is not formated yet (ext4)
I can do it manually at boot, but how can i format the partition so it's ready and mountable after flashing the device?
I know it's possible to do something like [this](https://github.com/OE4T/meta-tegra/discussions/858), but how can i do it when it's not a specific size (it needs to fill out remaining space)?
I also found [this thread](https://github.com/OE4T/meta-tegra/issues/143) that describes a way to use the initramfs to run a script (mkfs.ext4). Since I'm flashing with `.initrd-flash` already, do you have any pointers to how I can add a custom script that the initramfs will run when flashing?
Or maybe you've got an even better way of doing this?

**Reply by madisongh** [reply]
There are multiple ways to do this sort of thing. If you just need to have the partition formatted on first boot, and you're using systemd, then adding the `x-systemd.makefs` option to your `fstab` is probably the simplest approach.  The [systemd.mount](https://www.freedesktop.org/software/systemd/man/latest/systemd.mount.html#) man page describes that option.

If you have content you need to flash into that partition, you can use that first discussion you linked to, or use the `IMAGE_TEGRAFLASH_DATA` and `DATAFILE` variables, which we have support for in the flashing scripts.  You can use the `x-systemd.growfs` mount option to have the filesystem expanded to fill the partition completely at boot time.

And yes, you can customize the initrd-flash process to add in a formatting-only step for partitions.  There aren't any generic hooks in the scripts for this (something I'd like to add, at some point), but you could use a modified version of the [init-flash.sh](https://github.com/OE4T/meta-tegra/blob/master/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh).

**Reply by mickeyproc** [reply, marked as answer]
@madisongh 
The plan is to eventually use another init manager other than systemd, so I want to make sure this part is future proofed a bit.

I'm thinking of just calling a script from the `init-flash.sh` script that you mentioned. This init extra script is bbappended to the `tegra-flash-init.bb` recipe:
```
FILESEXTRAPATHS:prepend := "${THISDIR}/files:"

SRC_URI:append = "\
    file://custom_flash_init_extra.sh \
    "

RDEPENDS:${PN}:append = " util-linux-findfs"

EXTRA_INIT = "custom_flash_init_extra.sh"
```
The script i added just uses `mke2fs -t ext4` to format the custom added partitions as ext4 fs.
I've tried making it a bit generic so any custom stuff can be added to this `init_extra` script. One just needs to add the file to the `EXTRA_INIT` variable.

Maybe this should be a PR, unless you're thinking of a better way of doing it?:

``` 
diff --git a/recipes-bsp/tegra-binaries/tegra-helper-scripts/initrd-flash.sh b/recipes-bsp/tegra-binaries/tegra-helper-scripts/initrd-flash.sh
index 3eaa8834..ff1eb15c 100644
--- a/recipes-bsp/tegra-binaries/tegra-helper-scripts/initrd-flash.sh
+++ b/recipes-bsp/tegra-binaries/tegra-helper-scripts/initrd-flash.sh
@@ -365,7 +365,7 @@ generate_flash_package() {
        [ $EXTERNAL_ROOTFS_DRIVE -eq 0 -o $NO_INTERNAL_STORAGE -eq 1 ] || echo "erase-mmc" >> "$mnt/flashpkg/conf/command_sequence"
        echo "export-devices $ROOTFS_DEVICE" >> "$mnt/flashpkg/conf/command_sequence"
     fi
-
+    echo "extra" >> "$mnt/flashpkg/conf/command_sequence"
     echo "reboot" >> "$mnt/flashpkg/conf/command_sequence"
 
     unmount_and_release "$mnt" "$dev" || return 1
diff --git a/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh b/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh
index dd4ebc9c..c81a7ca0 100644
--- a/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh
+++ b/recipes-core/initrdscripts/tegra-flash-init/init-flash.sh
@@ -181,6 +181,13 @@ else
                    break
                done
                ;;
+           extra)
+               if [ -f "/init_extra" ]; then
+                   ./init_extra 2>&1 | tee /tmp/flashpkg/flashpkg/logs/custom-extra.log
+               else
+                   echo "No init_extra was found, ignoring" > /tmp/flashpkg/flashpkg/logs/custom-extra.log
+               fi
+               ;;
            reboot)
                reboot_type="$args"
                final_status="SUCCESS"
diff --git a/recipes-core/initrdscripts/tegra-flash-init_1.0.bb b/recipes-core/initrdscripts/tegra-flash-init_1.0.bb
index afa61201..1bc41cdf 100644
--- a/recipes-core/initrdscripts/tegra-flash-init_1.0.bb
+++ b/recipes-core/initrdscripts/tegra-flash-init_1.0.bb
@@ -12,8 +12,13 @@ COMPATIBLE_MACHINE = "(tegra)"
 
 S = "${WORKDIR}"
 
+EXTRA_INIT = ""
+
 do_install() {
     install -m 0755 ${WORKDIR}/init-flash.sh ${D}/init
+    if [ -f "${EXTRA_INIT}" ]; then
+        install -m 0755 ${WORKDIR}/${EXTRA_INIT} ${D}/init_extra
+    fi
     install -m 0555 -d ${D}/proc ${D}/sys
     install -m 0755 -d ${D}/dev ${D}/mnt ${D}/run ${D}/usr
     install -m 1777 -d ${D}/tmp
``` 

UPDATE:
I can see PR #1356 already does something alike what i was thinking, but my way is coincidentally more or less the exact way you described in a comment to the other PR. Made PR #1535 for review.
